### PR TITLE
Enhance type-driven optimisation in beam_type.erl

### PR DESCRIPTION
**Pull request by michalmuskala**
* kill type information only for affected registers in get_map_elements

* bs_get_utf* will produce integers of unicode range

  This optimises code created by Elixir compiler, where:

      <<x::utf8,_::binary>> when x in 1..10

  will compile the guard to

      is_integer(X) andalso X >= 1 andalso X =< 10

  This allows us to eliminate the is_integer check.

* bs_get_float will produce a float

* allow carrying type information over other bs instructions killing
  only the affected registers

* kill only x registers after call_fun and apply instructions

---

I didn't add tests for this since I wasn't exactly sure what to test here. I'm open to suggestions on how to test the introduced changes.

### CODE REVIEW

File: lib/compiler/src/beam_type.erl [hunk 1]
```diff
@@ -507,10 +512,39 @@ update({test,is_eq_exact,_,[Reg,{atom,_}=Atom]}, Ts) ->
...
+    Ts;
+update({bs_restore2,_,_}, Ts) ->
+    Ts;
+update({bs_context_to_binary,Dst}, Ts) ->
+    tdb_update([{Dst,kill}], Ts);
+update({test,bs_start_match2,_,_,_,Dst}, Ts) ->
+    tdb_update([{Dst,kill}], Ts);
+update({test,bs_get_binary2,_,_,_,Dst}, Ts) ->
+    tdb_update([{Dst,kill}], Ts);
+update({test,bs_get_float,_,_,_,Dst}, Ts) ->
```
Comment (bjorng) [inline lib/compiler/src/beam_type.erl]: Should be `bs_get_float2`, not `bs_get_float`.

**Comment by bjorng** [comment]
I can't think of any new tests to add. All of `beam_type` (except for the `bs_get_float` clause) is already covered. When you have fixed that little mistake, I will add the branch to our daily builds.

**Comment by michalmuskala** [comment]
@bjorng updated

**Comment by bjorng** [comment]
I found one more thing. In your handling of `get_map_elements`, source registers as well as destination registers will be killed. It will make a difference in this (contrived) example:

```
t(Bin, M) ->
    <<K:8>> = Bin,
    #{K:=Reg} = M,
    if
        is_integer(K) ->
            Reg;
        true ->
            not_integer
    end.
```
I have pushed a fixup commit suggesting how it can be fixed. If you like it, please squash it into your existing commit.

**Comment by michalmuskala** [comment]
That's definitely an improvement, thank you.

**Comment by michalmuskala** [comment]
There's one more optimisation I was meaning to do, but failed. I would be grateful for some tips on how to achieve that.

The problematic code is basically matching twice on the same map with the same pattern:
```erlang
test(#{test := foo} = Map) ->
    case Map of
        #{test := foo} = Map1 ->
            Map1#{test := bar};
        _ ->
            erlang:error({badarg, Map})
    end.
```

The generated assembly will unnecessarily match twice on the map and fail to eliminate the unreachable branch in the case:
```S
{function, test, 1, 2}.
  {label,1}.
    {line,[{location,"lib/test.erl",5}]}.
    {func_info,{atom,test},{atom,test},1}.
  {label,2}.
    {test,is_map,{f,1},[{x,0}]}.
    {get_map_elements,{f,1},{x,0},{list,[{atom,test},{x,1}]}}.
    {test,is_eq_exact,{f,1},[{x,1},{atom,foo}]}.
    {get_map_elements,{f,3},{x,0},{list,[{atom,bar},{x,2},{atom,test},{x,1}]}}.
    {test,is_eq_exact,{f,3},[{x,2},{atom,baz}]}.
    {test,is_eq_exact,{f,3},[{x,1},{atom,foo}]}.
    {line,[{location,"lib/test.erl",8}]}.
    {put_map_exact,{f,0},{x,0},{x,0},1,{list,[{atom,test},{atom,bar}]}}.
    return.
  {label,3}.
    {test_heap,3,1}.
    {put_tuple,2,{x,1}}.
    {put,{atom,badstruct}}.
    {put,{x,0}}.
    {move,{x,1},{x,0}}.
    {line,[{location,"lib/test.erl",10}]}.
    {call_ext,1,{extfunc,erlang,error,1}}.
```

The optimisation in this PR allows to eliminate unnecessary `is_map` tests (without the PR there would be an `is_map` test before each `get_map_elements`) but leaves duplicate extraction and value checks.

Similar cases of multiple matches on the same tuple are nicely optimised by the compiler, but maps are not optimised as well. The main problem is that I'm not sure when such an optimisation should happen - should it be at the core level in `sys_core_fold` or maybe somewhere later/earlier?

**Comment by bjorng** [comment]
The question of where to put an optimization is a complicated one which I ask myself all the time.

In this case, I suggest that you do it in `beam_type`. Look at the type `tuple_element` in the type database. You could implement a similar type called `map_element` or `map_elements`.

Since the implementation is not trivial, we would not want to include it unless that kind of code is used in real-world programs.

**Comment by michalmuskala** [comment]
Code like this is emitted by the Elixir compiler for structs, i.e.
```elixir
def test(%Type{} = type) do
  %Type{type | foo: :bar}
end
```

which is similar to a record operation:
```erlang
test(#type{} = Type) ->
  Type#type{foo = bar}
```

It might be possible to optimise code generation in the Elixir compiler, I will see what will be easier.

**Pull request merged**